cmake/FindAsio.cmake

```diff
@@ -0,0 +1,96 @@
+#[=======================================================================[.rst
+FindAsio
+--------
+
+FindModule for Asio and the associated library
+
+Imported Targets
+^^^^^^^^^^^^^^^^
+
+This module defines the :prop_tgt:`IMPORTED` target ``Asio::Asio``.
+
+Result Variables
+^^^^^^^^^^^^^^^^
+
+This module sets the following variables:
+
+``Asio_FOUND``
+  True, if the library was found.
+``Asio_VERSION``
+  Detected version of found Asio library.
+
+Cache variables
+^^^^^^^^^^^^^^^
+
+The following cache variables may also be set:
+
+``Asio_INCLUDE_DIR``
+  Directory containing ``asio.hpp``.
+
+Distributed under the MIT License, see accompanying LICENSE file or
+https://github.com/PatTheMav/cmake-finders/blob/master/LICENSE for details.
+(c) 2023 Trystan Mata
+
+#]=======================================================================]
+
+# cmake-format: off
+# cmake-lint: disable=C0103
+# cmake-lint: disable=C0301
+# cmake-format: on
+
+include(FindPackageHandleStandardArgs)
+
+find_package(PkgConfig QUIET)
+if(PKG_CONFIG_FOUND)
+  pkg_search_module(PC_Asio QUIET asio)
+endif()
+
+find_path(
+  Asio_INCLUDE_DIR
+  NAMES asio.hpp
+  HINTS ${PC_Asio_INCLUDE_DIRS}
+  PATHS /usr/include /usr/local/include
+  DOC "Asio include directory")
+
+if(PC_Asio_VERSION VERSION_GREATER 0)
+  set(Asio_VERSION ${PC_Asio_VERSION})
+elseif(EXISTS "${Asio_INCLUDE_DIR}/asio/version.hpp")
+  file(STRINGS "${Asio_INCLUDE_DIR}/asio/version.hpp" _version_string
+       REGEX "#define[ \t]+ASIO_VERSION[ \t]+[0-9]+[ \t]+\/\/[ \t][0-9]+\.[0-9]+\.[0-9]+")
+
+  string(REGEX REPLACE "#define[ \t]+ASIO_VERSION[ \t]+[0-9]+[ \t]+\/\/[ \t]([0-9]+\.[0-9]+\.[0-9]+)" "\\1"
+                       Asio_VERSION "${_version_string}")
+else()
+  if(NOT Asio_FIND_QUIETLY)
+    message(AUTHOR_WARNING "Failed to find Asio version.")
+  endif()
+  set(Asio_VERSION 0.0.0)
+endif()
+
+if(CMAKE_HOST_SYSTEM_NAME MATCHES "Darwin|Windows")
+  set(Asio_ERROR_REASON "Ensure that a Asio distribution is provided as part of CMAKE_PREFIX_PATH.")
+elseif(CMAKE_HOST_SYSTEM_NAME MATCHES "Linux|FreeBSD")
+  set(Asio_ERROR_REASON "Ensure Asio library is available in local include paths.")
```

Comment:
If it's a single library, this should either say "an Asio library" or "Asio libraries" otherwise.